Added tags model and collection

SaraMcCutcheon opened the pull request.

Commit:
Added tags model and collection

SaraMcCutcheon commented on php/Terminus/Helpers/InputHelper.php:
```diff
@@ -376,62 +376,60 @@ public function optional(array $arg_options = [])
-        'key'        => 'org',
-        'default'    => null,
-        'allow_none' => true,
+            'args'            => [],
```
This change prevents the only org present from being automatically chosen when there is just one.

SaraMcCutcheon commented on php/Terminus/Models/TerminusModel.php:
```diff
@@ -6,25 +6,25 @@
+    /**
+     * @var Request
+     */
+    public $request;
```
I would rather not make $request public. I've not any experience mocking a protected object. Any help?

ajbarry replied:
Yeah, ideally Models would be injected with RESTful datasource adapter but that's a bigger task. Perhaps `$__request` would belay the intent to keep this internal enough for now.

ajbarry replied:
Otherwise adding accessors around a private var would be the other way to go and set a custom mock object in the tests?

SaraMcCutcheon replied:
This has been worked around by applying RequestAwareTrait&Interface to TerminusModel and TerminusCollection.

SaraMcCutcheon commented on tests/unit_tests/Collections/TagsTest.php:
```diff
@@ -0,0 +1,77 @@
+        $tag_id = 'tag_id';
+        $this->tags->org_site_membership->site->id = 'site_uuid';
+        $this->tags->org_site_membership->organization->id = 'org_uuid';
+        $this->tags->request->expects($this->once())
```
This is where CI is failing - the tags request property expectations.

ajbarry commented on tests/features/site_tags.feature:
```diff
@@ -20,7 +20,7 @@ Feature: Tagging organizational sites
     Then I should get:
     """
-    {"tags":["testtag"]}
+    ["testtag"]
```
In lieu of a `Then I should get a JSON object with ...` step, I liked the old test a little better that asserts the whole structure of the JSON object returned. /shrug

We can add better JSON step matcher later that round-trips the output through JSON parsing to a PHP object that can be inspected.

SaraMcCutcheon replied:
I think this way is better than the old way. In other commands when the user asks for a list, we do not return a JSON object from the API, but an array.

SaraMcCutcheon commented:
I rebased this with the new TerminusModel/TerminusCollection using the @ronan's RequestAwareInterface/Trait. Moved Tag/Tags to 1.x models & collections, fixed unit tests.

SaraMcCutcheon commented on php/Terminus/Models/Site.php:
```diff
@@ -197,6 +153,21 @@ public function deleteBranch($branch)
+     * @param string[] $product_id The uuid for the product to deploy.
+     * @return Workflow
+     */
+    public function deployProduct($product_id)
```
Alphabetized this up to here.

bensheldon commented on php/Terminus/Collections/Tags.php:
```diff
@@ -0,0 +1,57 @@
+                $this->add((object)['id' => $tag_string,]);
+            }
+        }
+        return $this;
```
Why is this method called fetch? It seems like it's more of an `addMany` (or something) because it doesn't make an API request.

SaraMcCutcheon replied:
It is misnamed.

bensheldon commented on tests/unit_tests/new/Collections/TagsTest.php:
```diff
@@ -0,0 +1,66 @@
+    public function testFetch()
+    {
+        $data = ['tag1',];
+        $this->collection->fetch(compact('data'));
```
should there be an assertion here?

bensheldon reviewed (approved):
I'm a little confused by the semantics of `Tags#fetch` but not blocking. 👍
